tests/PHPStan/Analyser/data/TestDynamicReturnTypeExtensions.php

```diff
@@ -209,3 +217,107 @@ public function getTypeFromMethodCall(MethodReflection $methodReflection, Method
+ *
+ * Used in atk4/data repo.
+ */
+class Bug7344DynamicReturnTypeExtension implements DynamicMethodReturnTypeExtension
+{
+	public function getClass(): string
+	{
+		return \Bug7344\Model::class;
+	}
+
+	public function isMethodSupported(MethodReflection $methodReflection): bool
+	{
+		return true;
+	}
+
+	protected function unresolveMethodReflection(ResolvedMethodReflection $methodReflection): MethodReflection
+	{
+		$methodReflection = \Closure::bind(function () use ($methodReflection) { return $methodReflection->reflection; }, null, ResolvedMethodReflection::class)();
+		if (!$methodReflection instanceof ChangedTypeMethodReflection) {
+			throw new \Exception('Unexpected method reflection class: ' . get_class($methodReflection));
+		}
+
+		$methodReflection = \Closure::bind(function () use ($methodReflection) { return $methodReflection->reflection; }, null, ChangedTypeMethodReflection::class)();
+
+		return $methodReflection;
+	}
+
+	protected function resolveMethodReflection(MethodReflection $methodReflection, Type $calledOnType): MethodReflection
+	{
+		$resolver = (new CalledOnTypeUnresolvedMethodPrototypeReflection(
+			$methodReflection,
+			$methodReflection->getDeclaringClass(),
+			false,
+			$calledOnType
+		));
+
+		return $resolver->getTransformedMethod();
+	}
+
+	protected function reresolveMethodReflection(MethodReflection $methodReflection, Type $calledOnType): MethodReflection
+	{
+		if ($methodReflection instanceof UnionTypeMethodReflection) {
+			$methodReflection = new UnionTypeMethodReflection(
+				$methodReflection->getName(),
+				array_map(
+					function ($v) use ($calledOnType) { return $this->reresolveMethodReflection($v, $calledOnType); },
+					\Closure::bind(function () use ($methodReflection) { return $methodReflection->methods; }, null, UnionTypeMethodReflection::class)()
+				)
+			);
+		} else {
+			$methodReflection = $this->unresolveMethodReflection($methodReflection);
+			$methodReflection = $this->resolveMethodReflection($methodReflection, $calledOnType);
+		}
+
+		return $methodReflection;
+	}
+
+	protected function removeVirtualInterfacesFromType(Type $type): Type
+	{
+		if ($type instanceof IntersectionType) {
+			$types = [];
+			foreach ($type->getTypes() as $t) {
+				$t = $this->removeVirtualInterfacesFromType($t);
+				if (!$t instanceof NeverType) {
+					$types[] = $t;
+				}
+			}
+
+			return count($types) === 0 ? new NeverType() : TypeCombinator::intersect(...$types);
+		}
+
+		if ($type instanceof ObjectType && $type->isInstanceOf(\Bug7344\PhpdocTypeInterface::class)->yes()) {
+			return new NeverType();
+		}
+
+		return $type->traverse(\Closure::fromCallable([$this, 'removeVirtualInterfacesFromType']));
+	}
+
+	public function getTypeFromMethodCall(
+		MethodReflection $methodReflection,
+		MethodCall $methodCall,
+		Scope $scope
+	): Type {
+		// resolve static type and remove all virtual interfaces from it
+		if ($methodCall instanceof StaticCall) {
```

Comment:
This extension is too complicated. Can you just write a simple one just so we can verify it's actually executed? So we can see the method's return type isn't used but the one returned from an extension?

Reply:
this is the simplest impl. supporting unlimited nesting, please merge this test case as is

Reply:
The fixed issue can be demonstrated with a much simpler extension. Let's say you're writing an extension for a method that has `string` return type, the extension can just return `new IntegerType()` and we can assert that `int` has been used.

Reply:
the issue was limited nesting support of union/intersect

This test does support unlimited nesting - it removes virtual interface and reresolve `$this`/`static`. I do not think this two needed operations can be further simplified.

Reply:
Alright, had to do it myself :P https://github.com/phpstan/phpstan-src/pull/1380/commits/26d8fdebc35462948872156599892de607184827

I verified the test fails without the change in TypeUtils.